R/doubletFinder.R

```diff
@@ -87,13 +97,151 @@ doubletFinder <- function(seu, expected.doublets = 0, proportion.artificial = 0.
+  cells_both <- c(real.cells, paste0("X", 1:n_doublets))
+
+  # Make temporary loom file for data with simulated doublets
+  # Generate doublets
+  real_ind1 <- sample(1:n_real.cells, n_doublets, replace = TRUE)
+  real_ind2 <- sample(1:n_real.cells, n_doublets, replace = TRUE)
+  # If not chunked, this will load about half of the data into memory
+  # To conserve memory, get doublets by chunk
+  # As a rule of thumb, do calculation for 50 cells at a time
+  doublet_mat <- matrix(0, n_doublets, length(genes))
+  if (n_doublets <= 50) {
+    doublet_mat <-
+      (obj[["matrix"]][real_ind1,] + obj[["matrix"]][real_ind2,]) / 2
+  } else {
+    n_chunks <- floor(n_doublets / 50) # the last chunk is added separately
+    pb <- txtProgressBar(style = 3)
+    for (i in 1:n_chunks) {
+      ind1 <- 1 + (i - 1) * 50
+      ind2 <- i * 50
+      doublet_mat[ind1:ind2,] <-
+        (obj[["matrix"]][real_ind1[ind1:ind2],] + obj[["matrix"]][real_ind2[ind1:ind2],]) / 2
+      setTxtProgressBar(pb, i / n_chunks)
+    }
+    close(pb)
+    remainder <- n_chunks %% 50
+    if (remainder > 0) {
+      doublet_mat[(n_chunks * 50 + 1):n_doublets,] <-
+        (obj[["matrix"]][real_ind1[(ind2 + 1):n_doublets],] + obj[["matrix"]][real_ind2[(ind2 + 1):n_doublets],]) / 2
+    }
+  }
+  print("Writing artificial doublets to loom file")
+  doublets <- create_doublet_loom(n_doublets, doublet_mat, obj, overwrite = overwrite)
+  # Free up memory
+  rm(doublet_mat)
+  gc(verbose = FALSE)
+  # Combine real cells and doublets
+  print("Combining real cells and doublets")
+  loom_wdoublets <- loomR::combine(list(obj, doublets), "loom_wdoublets.loom",
+                                    overwrite = overwrite)
+  gc(verbose = FALSE)
+  ## Step 2: Pre-process real-artificial merged data using Seurat
+  print("Normalizing Seurat object...")
+  Seurat::NormalizeData(loom_wdoublets, overwrite = TRUE,
+                scale.factor = as.integer(hdf5r::h5attributes(obj[["layers/norm_data"]])$calc_params[1]),
+                chunk.size = chunk.size, chunk.dims = chunk.dims)
+  print("Finding variable genes...")
+  fvg_params <- as.numeric(hdf5r::h5attributes(obj[["row_attrs/var_genes"]])$calc_params[1:4])
+  Seurat::FindVariableGenes(loom_wdoublets, overwrite = TRUE,
+                            x.low.cutoff = fvg_params[1],
+                            x.high.cutooff = fvg_params[2],
+                            y.cutoff = fvg_params[3],
+                            y.high.cutoff = fvg_params[4],
+                            chunk.size = chunk.size)
+  print("Scaling data...")
+  Seurat::ScaleData(loom_wdoublets, display.progress = TRUE, overwrite = TRUE,
+                    chunk.size = chunk.size, chunk.dims = chunk.dims)
+  print("Running PCA...")
+  # Get the names of the highly variable genes
+  var_gene_inds <- loom_wdoublets[["row_attrs/var_genes"]][]
+  var_genes <- genes[var_gene_inds]
+  Seurat::RunPCA(loom_wdoublets, pcs.compute = pcs.compute, overwrite = TRUE,
```

Comment:
The number of pcs to compute will need to be more than 5.

Reply:
The default is 5, but in practice, for large loom datasets, I used 55 (depending on where the variance explained by the PCs plateau). But you're right, the default probably should be larger.